### DISCUSSION TITLE
exiting with a non-zero return code workaround broken

### DISCUSSION THREAD

**Discussion by mortax** [winglang/wing#5356] [Q&A] [answered]
this is related to my question about creating console output without a newline (#1772). i used the same technique to exit the program. i created `process-helper.js`:

```
exports.write = function(s) {
  return process.stdout.write(s);
};

exports.exit = function(rc) {
  return process.exit(rc);
};
```

adding the following in my program:

```
bring util;

class Process {
  extern "./process-helper.js" pub static write(s: str);
  extern "./process-helper.js" pub static exit(rc: num);
}
```

it's called as:

```
Process.exit(rc);
```

that worked perfectly through release `0.52`. starting with  `0.54`, the output isn't flushed flushed and my program no longer works. removing the call to `process.exit()` restores output (but breaks the return code workaround).

anyone know if this is a Wing or `node` issue?

**Comment by mortax** [comment]
perhaps you have an idea @MarkMcCulloh?

**Reply by MarkMcCulloh** [reply]
Yeah I'm sure this is related to the fact that preflight code is now executed in a worker thread, which handles process.exit differently. stdout is also different but it seems that works okay.

I'm curious what behavior you expect? Currently, I think `process.exit(0)` will simply immediately exit the preflight execution with no error, non-zero will cause an exception.

Also I should mention that process.exit is generally a risky idea because you may be skipping cleanup processes since the lifecycle of preflight execution is fully managed by the wing cli.

**Comment by mortax** [comment]
ah, a worker thread. that makes sense. what i really want to do i just unbuffer `stdout` (which is of course trivial in various scripting languages but seems to be nigh on impossible in `node.js`). all i really want to do is have my wing program return a value other than zero. `process.exit()` seemed the most straightforward way -- until now.

**Reply by MarkMcCulloh** [reply, marked as answer]
I think there is a bug preventing the exit code from propagating outside of the worker thread, so that part could be fixed.

Also I would suggest using `process.exitCode = <number>` instead to avoid the process terminating before the wing CLI can finish cleanup.

**Reply by mortax** [reply]
brilliant! i would finitely have done that first had i known. thanks a lot.

as an aside: it would be lovely to be able to control the exit code properly in Wing. i'd be willing to create the PR for that but even _seemingly_ simple tasks like that seem to balloon when i attempt them. that's what happening when i tried to create a `print`/`log` functionality sans `"\n"`. (i'd be happy to look with the proper shove in the right direction - i.e. a similar change - with tests - as an exemplar.)

**Comment by mortax** [comment]
@MarkMcCulloh i'm prepared to do a PR for this -- but it can't really be `exit()` since it won't end the program. it'll would have to be something like `util.exitCode(rc: num)` or `util.exitAs(rc: num)`, etc. what do you think?
